engine/src/main/java/org/terasology/rendering/nui/editor/NUIEditorContextMenuBuilder.java

```diff
@@ -0,0 +1,350 @@
+                addContextMenuListeners.forEach(UpdateListener::onAction);
+            }, node, true);
+            addLevel.addOption("String value", n -> {
+                n.addChild(new JsonTreeValue(null, "", JsonTreeValue.Type.VALUE));
+                addContextMenuListeners.forEach(UpdateListener::onAction);
+            }, node, true);
+        } else if (type == JsonTreeValue.Type.OBJECT) {
+            // Add a generic key/value pair addition option.
+            addLevel.addOption("Key/value pair", n -> {
+                n.addChild(new JsonTreeValue("", "", JsonTreeValue.Type.KEY_VALUE_PAIR));
+                addContextMenuListeners.forEach(UpdateListener::onAction);
+            }, node, true);
+
+            // Populate the menu with additional options.
+            if (node.hasChildWithKey("type")) {
+                populateWidgetContextMenu(node, addLevel);
+            } else if (node.getValue().getKey().equalsIgnoreCase("layoutInfo")) {
+                populateLayoutContextMenu(node, addLevel);
+            } else {
+                populateGenericContextMenu(node, addLevel);
+            }
+        }
+    }
+
+    private void populateWidgetContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        // Get the ClassMetadata of the widget with the given type.
+        String type = (String) node.getChildWithKey("type").getValue().getValue();
+        ClassMetadata<? extends UIWidget, ?> elementMetadata = nuiManager
+            .getWidgetMetadataLibrary()
+            .resolve(type, ModuleContext.getContext());
+
+        if (elementMetadata != null) {
+            for (FieldMetadata fieldMetadata : elementMetadata.getFields()) {
+                Field field = fieldMetadata.getField();
+                field.setAccessible(true);
+
+                // Add options for all @LayoutConfig-annotated fields.
+                if (field.isAnnotationPresent(LayoutConfig.class)) {
+                    String name = getNodeName(field);
+                    if (!node.hasChildWithKey(name)) {
+                        addLevel.addOption(name, n -> {
+                            try {
+                                createChild(name, node, field, elementMetadata);
+                            } catch (IllegalAccessException e) {
+                                return;
+                            }
+                            addContextMenuListeners.forEach(UpdateListener::onAction);
+                        }, node, true);
+                    }
+                }
+            }
+        } else {
+            logger.warn("Unknown widget type: {}", type);
+        }
+    }
+
+    private void populateLayoutContextMenu(JsonTree node, ContextMenuLevel addLevel) {
+        String type = "";
+        JsonTree currentNode = node;
+        ClassMetadata layoutMetadata = null;
+        Class layoutHintType = null;
+
+        // Iterate down the tree and attempt to find a UILayout node.
```

Comment:
This particular block can't be extracted into a method, since all of the variables declared above the comment are also used later. They could be wrapped within an inner class, but IMO that is unnecessarily complex.